xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+        trace_fmt!("Not enough data in the buffer!");
+        fail!("not_enough_data", ());
+    } else {
+        ()
+    };
+
+    let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
+    (
+        Buffer {
+            contents: buffer.contents >> length,
+            length: buffer.length - length
+        },
+        buffer.contents & mask
+    )
+}
+
+#[test]
+fn test_buffer_pop() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+    let (buffer, data) = buffer_pop(buffer, u32:16);
+    assert_eq(data, u32:0xBEEF);
+    assert_eq(buffer, Buffer { contents: u32:0xDEAD, length: u32:16 });
+    let (buffer, data) = buffer_pop(buffer, u32:16);
+    assert_eq(data, u32:0xDEAD);
+    assert_eq(buffer, Buffer { contents: u32:0, length: u32:0 });
+}
+
+// Pops `length` amount of data from a `buffer` if possible. The length of the
+// popped data is the same as the buffer size. Returns the status of the operation
+// and the updated buffer in a BufferResult structure, and the popped data.
+pub fn buffer_pop_checked<BSIZE: u32> (buffer: Buffer<BSIZE>, length: u32) -> (BufferResult<BSIZE>, bits[BSIZE]) {
+    if buffer_has_at_least(buffer, length) == false {
+        (
+            BufferResult { status: BufferStatus::NO_ENOUGH_DATA, buffer },
+            bits[BSIZE]:0
+        )
+    } else {
+        let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
+        (
+            BufferResult {
+                status: BufferStatus::OK,
+                buffer: Buffer {
+                    contents: buffer.contents >> length,
+                    length: buffer.length - length
+                }
+            },
+            buffer.contents & mask
+        )
+    }
+}
+
+#[test]
+fn test_buffer_pop_checked() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+
+    let (result1, data1) = buffer_pop_checked(buffer, u32:16);
+    assert_eq(result1, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xDEAD, length: u32:16 }
+    });
+    assert_eq(data1, u32:0xBEEF);
+
+    let (result2, data2) = buffer_pop_checked(result1.buffer, u32:16);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0, length: u32:0 }
+    });
+    assert_eq(data2, u32:0xDEAD);
+
+    let (result3, data3) = buffer_pop_checked(result2.buffer, u32:16);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_DATA,
+        buffer: result2.buffer
+    });
+    assert_eq(data3, u32:0);
+}
+
+// Pops `length` amount of data from a `buffer`. Returns the modified buffer
+// and the popped data. The length of the popped data is a function parameter.
+// It will fail if the buffer has no sufficient amount of data.
+// For calls that need better error handling, check `buffer_sized_pop_checked`.
+pub fn buffer_sized_pop<BSIZE: u32, WSIZE: u32> (buffer: Buffer<BSIZE>) -> (Buffer<BSIZE>, bits[WSIZE]) {
```

Comment:
maybe `fixed_pop` or `const_pop` would be a better name?

Reply:
I changed the name to `buffer_fixed_pop`